<a href="https://colab.research.google.com/github/HugoStigletz/Data-Analytics-Portfolio/blob/main/10_3_Training%2C_Validation%2C_%26_Testing_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Lesson 11.1.1 Activity

We know that height tends to run in families.  That is, tall parents tend to have children who grow up to be tall adults and vice versa.

But can we really predict how tall a child will grow up to be based only on how tall the parents are?  Does the prediction change if the child is a boy or a girl?  And is the same parent-child height relationship true for every family?

We’ll learn how to find answers to these questions during our study of linear regression.

Today, we'll begin the modeling process by splitting the data into training, validation, and test sets.



#Step 1: Download and save the `heights.csv` dataset from the class materials  

* Make a note of where you saved the file on your computer.

#Step 2: Upload the heights.csv dataset by running the following code block 

* When prompted, navigate to and select the `heights.csv` dataset where you saved it on your computer.
* This is a pretty large dataset, so it may take a little while.

In [ ]:
#Step 2

from google.colab import files
heights = files.upload()

Saving heights.csv to heights.csv


#Step 3: Import necessary packages
  * Import `pandas` as `pd`
  * `From sklearn.model_selection import train_test_split`

In [1]:
#Step 3
import pandas as pd
from sklearn.model_selection import train_test_split
import seaborn as sns
from statsmodels.formula.api import ols
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error as MSE

# Step 4: Create a Pandas DataFrame from the CSV file
* Name the DataFrame `heights`
* Print the first five observations of `heights`.  Note the kinds of data it contains

In [2]:
#Step 4
heights = pd.read_csv('/content/heights.csv')

heights.head(5)

Family  Father  Mother Gender  Height  Kids  MidParent
0       1    78.5    67.0      M    73.2     4      75.43
1       1    78.5    67.0      F    69.2     4      75.43
2       1    78.5    67.0      F    69.0     4      75.43
3       1    78.5    67.0      F    69.0     4      75.43
4       2    75.5    66.5      M    73.5     4      73.66

#Step 5: Print the shape of `heights`. 
* How many individuals were measured as part of the study?

In [5]:
#Step 5
heights.shape

(898, 7)

#Step 6: Split the data into the target variable and the feature of interest.
* We want to predict an adult child's height (`Height`) from the heights of the parents (`MidParent`).          
* `MidParent` is calculated as the average of the father's height plus a tiny bit more than the mother's height.
* Select the column `MidParent` from the `heights` DataFrame and name it `X`.  Use two sets of brackets around `MidParent` so X will be a DataFrame and not a Series (this is important for running the ML model later).  This has been done for you.
* Select the column `Height` from the `heights` DataFrame and name it `y`.  Make sure `y` is also a DataFrame and not a Series.





In [15]:
#Step 6
X = heights[['MidParent']]

y = heights[['Height']]

#Step 7: Split the data into a training/validation dataset and a test dataset.
* Use `train_test_split` from `sklearn.model_selection`.
* Name the X training/validation set `X_train_val` and the y training/validation set `y_train_val`.
* Name the X test set `X_test` and the y test set `y_test`.
* Set the `test_size = 0.25` and `random_state = 42`. 






In [34]:
#Step 7
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

# **## random_state=**

## We want our random numbers to be reproducable. If we want to randomly select things, the computer needs a starting point for how it's producing the random numbers. Where we set our random state it will determine where in the sequence our random number will start from. 

## An example is when you generate a new map in Civilization V. Your random state could be the type of land or coast line you'd like. Each new map will have a random land mass and coast but will follow that quideline. If I want a map of small islands, the random state will be small islands. I will always get small islands, but each new random map will have a different verzion of small islands. 

#Step 8: Print the shape of the X_train_val and X_test shape
* Make sure that the number of observations in the two datasets still add up to the original 898 records.





In [35]:
#Step 8
X_test.shape

(225, 1)

In [36]:
X_train_val

MidParent
377     68.730
357     69.270
723     68.060
306     69.938
464     70.140
..         ...
106     70.540
270     70.100
860     66.770
435     70.140
102     70.020

[673 rows x 1 columns]

In [37]:
X_test

MidParent
331     70.080
638     68.020
326     69.810
848     67.060
39      70.480
..         ...
101     70.020
199     69.520
851     67.060
346     69.398
299     70.100

[225 rows x 1 columns]

#Step 9: Split the training/validation dataset into a training set and validation set
* Use `train_test_split` from `sklearn.model_selection` to split `X_train_val` and `y_train_val` into `X_train`, `X_val`, `y_train` and `y_val`.
* Set the `test_size = 0.3333` (this will be the size of the validation set) and `random_state = 42`.





In [38]:
#Step 9
X_train, X_val, y_train, y_val =train_test_split(X_train_val, y_train_val, test_size=0.3333,random_state=42)

#Step 10: Print the shape of the X_train and X_val shape
* Make sure that the number of observations in the two datasets still add up to the 673 records in the `X_train_val` dataset.





In [39]:
print(X_test.shape)

(225, 1)


In [40]:
print(X_train)

     MidParent
760      69.18
93       70.02
881      66.56
34       70.48
440      70.41
..         ...
6        73.66
411      71.49
107      69.48
415      70.14
673      66.40

[448 rows x 1 columns]


In [41]:
print(X_val)

     MidParent
153      70.87
768      68.64
559      70.16
19       69.09
681      65.86
..         ...
542      66.90
457      70.39
189      69.79
269      70.10
663      66.94

[225 rows x 1 columns]


In [42]:
225 + 448

673